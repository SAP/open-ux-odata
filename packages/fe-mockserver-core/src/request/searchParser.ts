import { createToken, EmbeddedActionsParser, Lexer } from 'chevrotain';

// ----------------- Lexer -----------------

const OPEN = createToken({ name: 'OPEN', pattern: /(:?\(|%28)/ });
const CLOSE = createToken({ name: 'CLOSE', pattern: /(:?\)|%29)/ });
const QUOTE = createToken({ name: 'QUOTE', pattern: /(:?"|%22)/ });
const STRINGLITERAL = createToken({
    name: 'SimpleIdentifier',
    pattern: /(:?[^\\"\s]+|\\(:?[bfnrtv"\\/]|u[0-9a-fA-F]{4}))+/
});
// null, boolean, guid, dateTimeInOffset / dateValue / timeOfDay / decimalValue / doubleValue / singleValue / string / duration / enum / binary
const ANDOR = createToken({ name: 'AndOr', pattern: /(:?AND|OR)/ });
const WS = createToken({ name: 'Whitespace', pattern: /\s+/ });
const searchTokens = [OPEN, CLOSE, QUOTE, ANDOR, WS, STRINGLITERAL];

export const SearchLexer = new Lexer(searchTokens, {
    // Less position info tracked, reduces verbosity of the playground output.
    positionTracking: 'onlyStart'
});

// Parser
type CstRule<T> = (idxInCallingRule?: number, ...args: any[]) => T;
/**
 *
 */
export class SearchParser extends EmbeddedActionsParser {
    searchExpression: CstRule<string[]>;
    expression: CstRule<string[]>;
    constructor() {
        super(searchTokens, {
            recoveryEnabled: true
        });

        this.searchExpression = this.RULE('searchExpression', () => {
            return this.SUBRULE(this.expression);
        });

        this.expression = this.RULE('expression', () => {
            let searchTerms: any[] = [];
            const hasQuote = this.OPTION2(() => this.CONSUME(QUOTE));
            if (hasQuote) {
                // String literal including WS
                let stringLit = '';
                this.MANY_SEP({
                    SEP: WS,
                    DEF: () => {
                        const stringToken = this.CONSUME(STRINGLITERAL);
                        if (stringLit.length > 0) {
                            stringLit += ' ';
                        }
                        stringLit += stringToken.image;
                    }
                });
                searchTerms.push(stringLit);
                this.CONSUME2(QUOTE);
            } else {
                searchTerms.push(this.CONSUME(STRINGLITERAL).image);
            }
            this.OPTION3(() => {
                this.CONSUME(WS);
                const otherExpression = this.SUBRULE(this.expression);
                searchTerms = searchTerms.concat(otherExpression);
            });
            return searchTerms;
        });
        this.performSelfAnalysis();
    }
}
export function parseSearch(searchParameters: string | null): string[] {
    if (!searchParameters) {
        return [];
    }
    const parser = new SearchParser();
    const lexingResult = SearchLexer.tokenize(searchParameters);
    parser.input = lexingResult.tokens;
    const output = parser.searchExpression();

    if (parser.errors.length > 0) {
        parser.errors.forEach(console.error);
        throw new Error('Parsing errors detected');
    }
    return output;
}
