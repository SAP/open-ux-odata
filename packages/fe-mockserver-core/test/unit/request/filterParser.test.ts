import { parseFilter } from '../../../src/request/filterParser';

describe('Filter Parser', () => {
    test('can parse basic filtering', async () => {
        const basicEquals = parseFilter("Country_Code eq 'FR'");
        expect(basicEquals?.expressions[0].identifier).toBe('Country_Code');
        expect(basicEquals?.expressions[0].operator).toBe('eq');
        expect(basicEquals?.expressions[0].literal).toBe("'FR'");

        const basicGEquals = parseFilter('Country_Code ge 3');
        expect(basicGEquals?.expressions[0].identifier).toBe('Country_Code');
        expect(basicGEquals?.expressions[0].operator).toBe('ge');
        expect(basicGEquals?.expressions[0].literal).toBe('3');

        const navProp = parseFilter("Country/Code eq 'FR'");
        expect(navProp?.expressions[0].identifier).toBe('Country/Code');
        expect(navProp?.expressions[0].operator).toBe('eq');
        expect(navProp?.expressions[0].literal).toBe("'FR'");

        const multiNavProp = parseFilter("Country/Code/Toto eq 'FR'");
        expect(multiNavProp?.expressions[0].identifier).toBe('Country/Code/Toto');
        expect(multiNavProp?.expressions[0].operator).toBe('eq');
        expect(multiNavProp?.expressions[0].literal).toBe("'FR'");

        const equalWithSpace = parseFilter("Country_Code eq 'F R'");
        expect(equalWithSpace?.expressions[0].identifier).toBe('Country_Code');
        expect(equalWithSpace?.expressions[0].operator).toBe('eq');
        expect(equalWithSpace?.expressions[0].literal).toBe("'F R'");

        const dateEquals = parseFilter('ReleaseDate gt 2013-05-24');
        expect(dateEquals?.expressions[0].identifier).toBe('ReleaseDate');
        expect(dateEquals?.expressions[0].operator).toBe('gt');
        expect(dateEquals?.expressions[0].literal).toBe('2013-05-24');

        const dateBetween = parseFilter(
            '((SitnInstceCreatedAtDateTime gt 2022-05-15T22:00:00.000Z and SitnInstceCreatedAtDateTime lt 2022-05-16T21:59:59.000Z))'
        );
        expect(dateBetween?.expressions[0].identifier).toBe('SitnInstceCreatedAtDateTime');
        expect(dateBetween?.expressions[0].operator).toBe('gt');
        expect(dateBetween?.expressions[0].literal).toBe('2022-05-15T22:00:00.000Z');
        expect(dateBetween?.expressions[1].identifier).toBe('SitnInstceCreatedAtDateTime');
        expect(dateBetween?.expressions[1].operator).toBe('lt');
        expect(dateBetween?.expressions[1].literal).toBe('2022-05-16T21:59:59.000Z');

        const dateBetweenTZ = parseFilter(
            'DateTime ge 2022-05-16T00:00:00+02:00 and DateTime le 2022-05-22T23:59:59+02:00'
        );
        expect(dateBetweenTZ?.expressions[0].identifier).toBe('DateTime');
        expect(dateBetweenTZ?.expressions[0].operator).toBe('ge');
        expect(dateBetweenTZ?.expressions[0].literal).toBe('2022-05-16T00:00:00+02:00');
        expect(dateBetweenTZ?.expressions[1].identifier).toBe('DateTime');
        expect(dateBetweenTZ?.expressions[1].operator).toBe('le');
        expect(dateBetweenTZ?.expressions[1].literal).toBe('2022-05-22T23:59:59+02:00');

        const dateBetweenTZPlus = parseFilter(
            '(IsActiveEntity eq false or SiblingEntity/IsActiveEntity eq null) and DateTime ge 2022-05-23T00:00:00+02:00 and DateTime le 2022-05-29T23:59:59+02:00'
        );
        expect(dateBetweenTZPlus?.expressions[1].identifier).toBe('DateTime');
        expect(dateBetweenTZPlus?.expressions[1].operator).toBe('ge');
        expect(dateBetweenTZPlus?.expressions[1].literal).toBe('2022-05-23T00:00:00+02:00');
        expect(dateBetweenTZPlus?.expressions[2].identifier).toBe('DateTime');
        expect(dateBetweenTZPlus?.expressions[2].operator).toBe('le');
        expect(dateBetweenTZPlus?.expressions[2].literal).toBe('2022-05-29T23:59:59+02:00');

        const parenthesis = parseFilter("(Country_Code eq 'FR')");
        expect(parenthesis?.expressions[0].identifier).toBe('Country_Code');
        expect(parenthesis?.expressions[0].operator).toBe('eq');
        expect(parenthesis?.expressions[0].literal).toBe("'FR'");

        const guidParser = parseFilter(
            'sourceObject_ID eq ee1a9172-f3c3-47ce-b0f7-dd28c740210c or triggerObject_ID eq 837bc2bf-2e4b-4b2c-8c6c-9a9330c62400'
        );
        expect(guidParser?.expressions[0].identifier).toBe('sourceObject_ID');
        expect(guidParser?.expressions[0].operator).toBe('eq');
        expect(guidParser?.expressions[0].literal).toBe('ee1a9172-f3c3-47ce-b0f7-dd28c740210c');
    });

    test('can deal with AND / OR operator', () => {
        const andOrPrecedence = parseFilter("IsHot eq 'Yo' and IsHot eq true or PeopleCount gt 3");
        expect(andOrPrecedence?.operator).toBe('OR');
        expect(andOrPrecedence?.expressions.length).toBe(2);
        expect(andOrPrecedence?.expressions[0].expressions.length).toBe(2);
        expect(andOrPrecedence?.expressions[0].operator).toBe('AND');

        const andOrParenthesis = parseFilter('(IsHot eq false and IsHot eq true) or PeopleCount gt 3');
        expect(andOrParenthesis?.operator).toBe('OR');
        expect(andOrParenthesis?.expressions.length).toBe(2);
        expect(andOrParenthesis?.expressions[0].expressions.length).toBe(2);
        expect(andOrParenthesis?.expressions[0].operator).toBe('AND');

        const andOrParenthesis2 = parseFilter('IsHot eq false and (IsHot eq true or PeopleCount gt 3)');
        expect(andOrParenthesis2?.operator).toBe('AND');
        expect(andOrParenthesis2?.expressions.length).toBe(2);
        expect(andOrParenthesis2?.expressions[1].expressions.length).toBe(2);
        expect(andOrParenthesis2?.expressions[1].operator).toBe('OR');

        const andOrNoParenthesis = parseFilter('PeopleCount gt 3 or IsHot eq true and IsHot eq false');
        expect(andOrNoParenthesis?.operator).toBe('OR');
        expect(andOrNoParenthesis?.expressions.length).toBe(2);
        expect(andOrNoParenthesis?.expressions[1].expressions.length).toBe(2);
        expect(andOrNoParenthesis?.expressions[1].operator).toBe('AND');

        const dualParenthesisGroup = parseFilter(
            '(PeopleCount gt 3 or IsHot eq true) and (IsHot eq false or PeopleCount lt 5)'
        );
        expect(dualParenthesisGroup?.operator).toBe('AND');
        expect(dualParenthesisGroup?.expressions.length).toBe(2);
        expect(dualParenthesisGroup?.expressions[1].expressions.length).toBe(2);
        expect(dualParenthesisGroup?.expressions[1].operator).toBe('OR');
        expect(dualParenthesisGroup?.expressions[0].expressions.length).toBe(2);
        expect(dualParenthesisGroup?.expressions[0].operator).toBe('OR');
    });

    test('can deal with parenthesis', () => {
        const dualParenthesisGroup2 = parseFilter(
            '((PeopleCount gt 3 or IsHot eq true) and (IsHot eq false or PeopleCount lt 5))'
        );
        expect(dualParenthesisGroup2?.operator).toBe('AND');
        expect(dualParenthesisGroup2?.expressions.length).toBe(2);
        expect(dualParenthesisGroup2?.expressions[1].expressions.length).toBe(2);
        expect(dualParenthesisGroup2?.expressions[1].operator).toBe('OR');
        expect(dualParenthesisGroup2?.expressions[0].expressions.length).toBe(2);
        expect(dualParenthesisGroup2?.expressions[0].operator).toBe('OR');

        const dualParenthesisGroup3 = parseFilter(
            '((PeopleCount gt 3 or IsHot eq true) and PeopleCount lt 5) or PeopleCount ne 70'
        );
        expect(dualParenthesisGroup3?.operator).toBe('OR');
        expect(dualParenthesisGroup3?.expressions.length).toBe(2);
        expect(dualParenthesisGroup3?.expressions[0].expressions.length).toBe(2);
        expect(dualParenthesisGroup3?.expressions[0].operator).toBe('AND');
        expect(dualParenthesisGroup3?.expressions[0].expressions[0].expressions.length).toBe(2);
        expect(dualParenthesisGroup3?.expressions[0].expressions[0].operator).toBe('OR');
        expect(dualParenthesisGroup3?.expressions[1].literal).toBe('70');
        expect(dualParenthesisGroup3?.expressions[1].identifier).toBe('PeopleCount');
    });

    test('can deal with function calls', () => {
        // const toLowerCall = parseFilter("tolower(Country_Code) eq 'de'");
        // expect(toLowerCall.expressions[0].identifier.method).toBe('tolower');
        // expect(toLowerCall.expressions[0].identifier.methodArgs[0]).toBe('Country_Code');
        // expect(toLowerCall.expressions[0].operator).toBe('eq');
        // expect(toLowerCall.expressions[0].literal).toBe("'de'");
        //
        // const lengthCall = parseFilter('length(Country_Code) eq 5');
        // expect(lengthCall.expressions[0].identifier.method).toBe('length');
        // expect(lengthCall.expressions[0].identifier.methodArgs[0]).toBe('Country_Code');
        // expect(lengthCall.expressions[0].operator).toBe('eq');
        // expect(lengthCall.expressions[0].literal).toBe('5');
        //
        // const callsOnBothSide = parseFilter("tolower(Country_Code) eq tolower('de')");
        // expect(callsOnBothSide.expressions[0].identifier.method).toBe('tolower');
        // expect(callsOnBothSide.expressions[0].identifier.methodArgs[0]).toBe('Country_Code');
        // expect(callsOnBothSide.expressions[0].operator).toBe('eq');
        // expect(callsOnBothSide.expressions[0].literal.method).toBe('tolower');
        // expect(callsOnBothSide.expressions[0].literal.methodArgs[0]).toBe("'de'");
        //
        // const complexToLowerCall = parseFilter("(tolower(Country/Code) eq 'de' or Country_Code eq 'fr')");
        // expect(complexToLowerCall.expressions.length).toBe(2);
        // expect(complexToLowerCall.expressions[0].identifier.method).toBe('tolower');
        // expect(complexToLowerCall.expressions[0].identifier.methodArgs[0]).toBe('Country/Code');
        // expect(complexToLowerCall.expressions[0].operator).toBe('eq');
        // expect(complexToLowerCall.expressions[0].literal).toBe("'de'");
        // expect(complexToLowerCall.expressions[1].identifier).toBe('Country_Code');
        // expect(complexToLowerCall.expressions[1].operator).toBe('eq');
        // expect(complexToLowerCall.expressions[1].literal).toBe("'fr'");

        const multipleOredCall = parseFilter(
            "contains(externalId,'123') or contains(namespace,'123') or contains(name,'123')"
        );
        expect(multipleOredCall?.expressions.length).toBe(3);
        expect(typeof multipleOredCall?.expressions[0].identifier).toBe('object');
        if (
            typeof multipleOredCall?.expressions[0].identifier === 'object' &&
            multipleOredCall?.expressions[0].identifier?.type === 'method'
        ) {
            expect(multipleOredCall?.expressions[0].identifier?.method).toBe('contains');
            expect(multipleOredCall?.expressions[0].identifier?.methodArgs[0]).toBe('externalId');
            expect(multipleOredCall?.expressions[0].identifier?.methodArgs[1]).toBe("'123'");
        }
    });

    test('can deal with complex function calls', () => {
        const complexFunction = parseFilter("startswith(CompanyName,'Futterkiste')");
        expect(typeof complexFunction?.expressions[0].identifier).toBe('object');
        if (
            typeof complexFunction?.expressions[0].identifier === 'object' &&
            complexFunction?.expressions[0].identifier?.type === 'method'
        ) {
            expect(complexFunction?.expressions[0].identifier?.method).toBe('startswith');
            expect(complexFunction?.expressions[0].identifier?.methodArgs[0]).toBe('CompanyName');
            expect(complexFunction?.expressions[0].identifier?.methodArgs[1]).toBe("'Futterkiste'");
        }

        const complexFunctionWithNavProp = parseFilter("startswith(Company/Name,'Futterkiste')");
        expect(typeof complexFunctionWithNavProp?.expressions[0].identifier).toBe('object');
        if (
            typeof complexFunctionWithNavProp?.expressions[0].identifier === 'object' &&
            complexFunctionWithNavProp?.expressions[0].identifier?.type === 'method'
        ) {
            expect(complexFunctionWithNavProp?.expressions[0]?.identifier.method).toBe('startswith');
            expect(complexFunctionWithNavProp?.expressions[0]?.identifier.methodArgs[0]).toBe('Company/Name');
            expect(complexFunctionWithNavProp?.expressions[0]?.identifier.methodArgs[1]).toBe("'Futterkiste'");
        }

        const complexFunctionEquality = parseFilter("cast(ConfigurationStatus, Edm.String) eq 'COMPLETE'");
        expect(typeof complexFunctionEquality?.expressions[0].identifier).toBe('object');
        if (
            typeof complexFunctionEquality?.expressions[0].identifier === 'object' &&
            complexFunctionEquality?.expressions[0].identifier?.type === 'method'
        ) {
            expect(complexFunctionEquality?.expressions[0].identifier?.method).toBe('cast');
            expect(complexFunctionEquality?.expressions[0].identifier?.methodArgs[0]).toBe('ConfigurationStatus');
            expect(complexFunctionEquality?.expressions[0].identifier?.methodArgs[1]).toBe('Edm.String');
            expect(complexFunctionEquality?.expressions[0].operator).toBe('eq');
            expect(complexFunctionEquality?.expressions[0].literal).toBe("'COMPLETE'");
        }
    });

    test('can deal with nested function calls', () => {
        const nested = parseFilter("contains(trim(tolower(Description)), 'port')");
        expect((nested?.expressions[0].identifier as any)?.type).toBe('method');
        expect(nested).toMatchSnapshot();

        const nestedEverywhere = parseFilter("contains(trim(tolower(Description)), tolower('Port'))");
        expect((nestedEverywhere?.expressions[0].identifier as any)?.type).toBe('method');
        expect(nestedEverywhere).toMatchSnapshot();

        const nestedComplex = parseFilter("contains(trim(tolower(cast(Description, Edm.String))), 'port')");
        expect((nested?.expressions[0].identifier as any)?.type).toBe('method');
        expect(nestedComplex).toMatchSnapshot();
    });

    test('can deal with lambda', () => {
        const lamba = parseFilter("CountryCodes/any(ent:ent eq 'GBR')");
        expect(typeof lamba?.expressions[0].identifier).toBe('object');
        if (
            typeof lamba?.expressions[0].identifier === 'object' &&
            lamba?.expressions[0].identifier?.type === 'lambda'
        ) {
            expect(lamba?.expressions[0].identifier?.type).toBe('lambda');
            expect(lamba?.expressions[0].identifier?.operator).toBe('ANY');
            expect(lamba?.expressions[0].identifier?.key).toBe('ent');
            expect(lamba?.expressions[0].identifier?.expression.expressions.length).toBe(1);
            expect(lamba?.expressions[0].identifier?.expression.expressions[0].operator).toBe('eq');
            expect(lamba?.expressions[0].identifier?.expression.expressions[0].literal).toBe("'GBR'");
            expect(lamba?.expressions[0].identifier?.expression.expressions[0].identifier).toBe('ent');
            expect(lamba?.expressions[0].identifier?.target).toBe('CountryCodes');
        }

        const notLamba = parseFilter('(allPolicies eq true)');
        expect(notLamba?.expressions[0].operator).toBe('eq');
        expect(notLamba?.expressions[0].identifier).toBe('allPolicies');
        expect(notLamba?.expressions[0].literal).toBe('true');
    });

    test('can deal with empty values', () => {
        const emptyValues = parseFilter(
            "Customer eq '1' and CompanyCode eq '0001' and Currency eq 'EUR' and SalesOrganization eq '1010' and DistributionChannel eq '01' and Division eq '01' and PartnerCounter eq '' and IBAN eq '' and CardNumber eq '2922870' and PaymentCardType eq ''"
        );
        expect(emptyValues).toMatchSnapshot();
    });

    test('it can deal with v2 specific guid', () => {
        const v2Guid = parseFilter('ProdCmplncSrvcReqFailureUUID eq guid%2700000000-0000-0000-0002-000000000000%27');
        expect(v2Guid).toMatchSnapshot();
    });

    test('v2 complex filter', () => {
        const v2ComplexFilter2 = parseFilter(
            "(AdvncdPlngDisplayHorizonDte ge datetime'2021-01-01T00:00:00' and AdvncdPlngDisplayHorizonDte le datetime'2021-12-31T00:00:00')"
        );
        expect(v2ComplexFilter2).toMatchSnapshot();
        const v2ComplexFilter3 = parseFilter(
            "(AdvncdPlngResourceName eq 'AdvncdPlngResourceName 1' or AdvncdPlngResourceName eq 'AdvncdPlngResourceName 10' or AdvncdPlngResourceName eq 'AdvncdPlngResourceName 2' or AdvncdPlngResourceName eq 'AdvncdPlngResourceName 3' or AdvncdPlngResourceName eq 'AdvncdPlngResourceName 4' or AdvncdPlngResourceName eq 'AdvncdPlngResourceName 5' or AdvncdPlngResourceName eq 'AdvncdPlngResourceName 6' or AdvncdPlngResourceName eq 'AdvncdPlngResourceName 7' or AdvncdPlngResourceName eq 'AdvncdPlngResourceName 8' or AdvncdPlngResourceName eq 'AdvncdPlngResourceName 9')"
        );
        expect(v2ComplexFilter3).toMatchSnapshot();
        const v2ComplexFilter4 = parseFilter("SimulationSessionUUID eq guid'd535d9a2-fdea-4e1e-84e6-2c135fcbcd08'");
        expect(v2ComplexFilter4).toMatchSnapshot();
        const v2ComplexFilter = parseFilter(
            '(' +
                "(AdvncdPlngDisplayHorizonDte ge datetime'2021-01-01T00:00:00' and AdvncdPlngDisplayHorizonDte le datetime'2021-12-31T00:00:00') " +
                'and ' +
                "(AdvncdPlngResourceName eq 'AdvncdPlngResourceName 1' or AdvncdPlngResourceName eq 'AdvncdPlngResourceName 10' or AdvncdPlngResourceName eq 'AdvncdPlngResourceName 2' or AdvncdPlngResourceName eq 'AdvncdPlngResourceName 3' or AdvncdPlngResourceName eq 'AdvncdPlngResourceName 4' or AdvncdPlngResourceName eq 'AdvncdPlngResourceName 5' or AdvncdPlngResourceName eq 'AdvncdPlngResourceName 6' or AdvncdPlngResourceName eq 'AdvncdPlngResourceName 7' or AdvncdPlngResourceName eq 'AdvncdPlngResourceName 8' or AdvncdPlngResourceName eq 'AdvncdPlngResourceName 9') and PlanningVersionExternal eq '000') " +
                "and SimulationSessionUUID eq guid'd535d9a2-fdea-4e1e-84e6-2c135fcbcd08'"
        );
        expect(v2ComplexFilter).toMatchSnapshot();
    });

    test('v4 complex filter', () => {
        const v4ComplexLambda = parseFilter(
            "((_Trigger/any(t:t/_Attribute/any(a0:a0/SitnInstceAttribName eq 'CAMASSRUNDATE' and a0/_AttributeValue/any(a1:a1/SitnInstceAttribValue ge '20220601') and a0/_AttributeValue/any(a1:a1/SitnInstceAttribValue le '20220601'))))) and (SitnBaseTemplateID eq 'FICA_MASSACT_MSG_MONITOR')"
        );
        expect(v4ComplexLambda).toMatchSnapshot();
    });

    test('v2 alternative filter', () => {
        const v2ComplexFilter = parseFilter(
            "(Product eq '1234' and SalesOrganization eq '0DE1') and (ValidFrom le datetime'2022-10-07T06:52:24.189' and ValidTo ge datetime'2022-10-07T06:52:24.189')"
        );
        expect(v2ComplexFilter).toMatchSnapshot();
    });
});
