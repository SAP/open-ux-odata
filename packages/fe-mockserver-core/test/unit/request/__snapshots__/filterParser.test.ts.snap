// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter Parser can deal with empty values 1`] = `
{
  "expressions": [
    {
      "identifier": "Customer",
      "literal": "'1'",
      "operator": "eq",
    },
    {
      "identifier": "CompanyCode",
      "literal": "'0001'",
      "operator": "eq",
    },
    {
      "identifier": "Currency",
      "literal": "'EUR'",
      "operator": "eq",
    },
    {
      "identifier": "SalesOrganization",
      "literal": "'1010'",
      "operator": "eq",
    },
    {
      "identifier": "DistributionChannel",
      "literal": "'01'",
      "operator": "eq",
    },
    {
      "identifier": "Division",
      "literal": "'01'",
      "operator": "eq",
    },
    {
      "identifier": "PartnerCounter",
      "literal": "''",
      "operator": "eq",
    },
    {
      "identifier": "IBAN",
      "literal": "''",
      "operator": "eq",
    },
    {
      "identifier": "CardNumber",
      "literal": "'2922870'",
      "operator": "eq",
    },
    {
      "identifier": "PaymentCardType",
      "literal": "''",
      "operator": "eq",
    },
  ],
  "operator": "AND",
}
`;

exports[`Filter Parser can deal with nested function calls 1`] = `
{
  "expressions": [
    {
      "identifier": {
        "method": "contains",
        "methodArgs": [
          {
            "method": "trim",
            "methodArgs": [
              {
                "method": "tolower",
                "methodArgs": [
                  "Description",
                ],
                "type": "method",
              },
            ],
            "type": "method",
          },
          "'port'",
        ],
        "type": "method",
      },
    },
  ],
}
`;

exports[`Filter Parser can deal with nested function calls 2`] = `
{
  "expressions": [
    {
      "identifier": {
        "method": "contains",
        "methodArgs": [
          {
            "method": "trim",
            "methodArgs": [
              {
                "method": "tolower",
                "methodArgs": [
                  "Description",
                ],
                "type": "method",
              },
            ],
            "type": "method",
          },
          {
            "method": "tolower",
            "methodArgs": [
              "'Port'",
            ],
            "type": "method",
          },
        ],
        "type": "method",
      },
    },
  ],
}
`;

exports[`Filter Parser can deal with nested function calls 3`] = `
{
  "expressions": [
    {
      "identifier": {
        "method": "contains",
        "methodArgs": [
          {
            "method": "trim",
            "methodArgs": [
              {
                "method": "tolower",
                "methodArgs": [
                  {
                    "method": "cast",
                    "methodArgs": [
                      "Description",
                      "Edm.String",
                    ],
                    "type": "method",
                  },
                ],
                "type": "method",
              },
            ],
            "type": "method",
          },
          "'port'",
        ],
        "type": "method",
      },
    },
  ],
}
`;

exports[`Filter Parser it can deal with v2 specific guid 1`] = `
{
  "expressions": [
    {
      "identifier": "ProdCmplncSrvcReqFailureUUID",
      "literal": "guid%2700000000-0000-0000-0002-000000000000%27",
      "operator": "eq",
    },
  ],
}
`;

exports[`Filter Parser v2 alternative filter 1`] = `
{
  "expressions": [
    {
      "expressions": [
        {
          "identifier": "Product",
          "literal": "'1234'",
          "operator": "eq",
        },
        {
          "identifier": "SalesOrganization",
          "literal": "'0DE1'",
          "operator": "eq",
        },
      ],
      "isGroup": true,
      "operator": "AND",
    },
    {
      "identifier": "ValidFrom",
      "literal": "datetime'2022-10-07T06:52:24.189'",
      "operator": "le",
    },
    {
      "identifier": "ValidTo",
      "literal": "datetime'2022-10-07T06:52:24.189'",
      "operator": "ge",
    },
  ],
  "operator": "AND",
}
`;

exports[`Filter Parser v2 complex filter 1`] = `
{
  "expressions": [
    {
      "identifier": "AdvncdPlngDisplayHorizonDte",
      "literal": "datetime'2021-01-01T00:00:00'",
      "operator": "ge",
    },
    {
      "identifier": "AdvncdPlngDisplayHorizonDte",
      "literal": "datetime'2021-12-31T00:00:00'",
      "operator": "le",
    },
  ],
  "isGroup": true,
  "operator": "AND",
}
`;

exports[`Filter Parser v2 complex filter 2`] = `
{
  "expressions": [
    {
      "identifier": "AdvncdPlngResourceName",
      "literal": "'AdvncdPlngResourceName 1'",
      "operator": "eq",
    },
    {
      "identifier": "AdvncdPlngResourceName",
      "literal": "'AdvncdPlngResourceName 10'",
      "operator": "eq",
    },
    {
      "identifier": "AdvncdPlngResourceName",
      "literal": "'AdvncdPlngResourceName 2'",
      "operator": "eq",
    },
    {
      "identifier": "AdvncdPlngResourceName",
      "literal": "'AdvncdPlngResourceName 3'",
      "operator": "eq",
    },
    {
      "identifier": "AdvncdPlngResourceName",
      "literal": "'AdvncdPlngResourceName 4'",
      "operator": "eq",
    },
    {
      "identifier": "AdvncdPlngResourceName",
      "literal": "'AdvncdPlngResourceName 5'",
      "operator": "eq",
    },
    {
      "identifier": "AdvncdPlngResourceName",
      "literal": "'AdvncdPlngResourceName 6'",
      "operator": "eq",
    },
    {
      "identifier": "AdvncdPlngResourceName",
      "literal": "'AdvncdPlngResourceName 7'",
      "operator": "eq",
    },
    {
      "identifier": "AdvncdPlngResourceName",
      "literal": "'AdvncdPlngResourceName 8'",
      "operator": "eq",
    },
    {
      "identifier": "AdvncdPlngResourceName",
      "literal": "'AdvncdPlngResourceName 9'",
      "operator": "eq",
    },
  ],
  "isGroup": true,
  "operator": "OR",
}
`;

exports[`Filter Parser v2 complex filter 3`] = `
{
  "expressions": [
    {
      "identifier": "SimulationSessionUUID",
      "literal": "guid'd535d9a2-fdea-4e1e-84e6-2c135fcbcd08'",
      "operator": "eq",
    },
  ],
}
`;

exports[`Filter Parser v2 complex filter 4`] = `
{
  "expressions": [
    {
      "expressions": [
        {
          "expressions": [
            {
              "identifier": "AdvncdPlngDisplayHorizonDte",
              "literal": "datetime'2021-01-01T00:00:00'",
              "operator": "ge",
            },
            {
              "identifier": "AdvncdPlngDisplayHorizonDte",
              "literal": "datetime'2021-12-31T00:00:00'",
              "operator": "le",
            },
          ],
          "isGroup": true,
          "operator": "AND",
        },
        {
          "expressions": [
            {
              "identifier": "AdvncdPlngResourceName",
              "literal": "'AdvncdPlngResourceName 1'",
              "operator": "eq",
            },
            {
              "identifier": "AdvncdPlngResourceName",
              "literal": "'AdvncdPlngResourceName 10'",
              "operator": "eq",
            },
            {
              "identifier": "AdvncdPlngResourceName",
              "literal": "'AdvncdPlngResourceName 2'",
              "operator": "eq",
            },
            {
              "identifier": "AdvncdPlngResourceName",
              "literal": "'AdvncdPlngResourceName 3'",
              "operator": "eq",
            },
            {
              "identifier": "AdvncdPlngResourceName",
              "literal": "'AdvncdPlngResourceName 4'",
              "operator": "eq",
            },
            {
              "identifier": "AdvncdPlngResourceName",
              "literal": "'AdvncdPlngResourceName 5'",
              "operator": "eq",
            },
            {
              "identifier": "AdvncdPlngResourceName",
              "literal": "'AdvncdPlngResourceName 6'",
              "operator": "eq",
            },
            {
              "identifier": "AdvncdPlngResourceName",
              "literal": "'AdvncdPlngResourceName 7'",
              "operator": "eq",
            },
            {
              "identifier": "AdvncdPlngResourceName",
              "literal": "'AdvncdPlngResourceName 8'",
              "operator": "eq",
            },
            {
              "identifier": "AdvncdPlngResourceName",
              "literal": "'AdvncdPlngResourceName 9'",
              "operator": "eq",
            },
          ],
          "isGroup": true,
          "operator": "OR",
        },
        {
          "identifier": "PlanningVersionExternal",
          "literal": "'000'",
          "operator": "eq",
        },
      ],
      "isGroup": true,
      "operator": "AND",
    },
    {
      "identifier": "SimulationSessionUUID",
      "literal": "guid'd535d9a2-fdea-4e1e-84e6-2c135fcbcd08'",
      "operator": "eq",
    },
  ],
  "operator": "AND",
}
`;

exports[`Filter Parser v4 complex filter 1`] = `
{
  "expressions": [
    {
      "expressions": [
        {
          "identifier": {
            "expression": {
              "expressions": [
                {
                  "identifier": {
                    "expression": {
                      "expressions": [
                        {
                          "identifier": "a0/SitnInstceAttribName",
                          "literal": "'CAMASSRUNDATE'",
                          "operator": "eq",
                        },
                        {
                          "identifier": {
                            "expression": {
                              "expressions": [
                                {
                                  "identifier": "a1/SitnInstceAttribValue",
                                  "literal": "'20220601'",
                                  "operator": "ge",
                                },
                              ],
                            },
                            "key": "a1",
                            "operator": "ANY",
                            "target": "a0/_AttributeValue",
                            "type": "lambda",
                          },
                        },
                        {
                          "identifier": {
                            "expression": {
                              "expressions": [
                                {
                                  "identifier": "a1/SitnInstceAttribValue",
                                  "literal": "'20220601'",
                                  "operator": "le",
                                },
                              ],
                            },
                            "key": "a1",
                            "operator": "ANY",
                            "target": "a0/_AttributeValue",
                            "type": "lambda",
                          },
                        },
                      ],
                      "operator": "AND",
                    },
                    "key": "a0",
                    "operator": "ANY",
                    "target": "t/_Attribute",
                    "type": "lambda",
                  },
                },
              ],
            },
            "key": "t",
            "operator": "ANY",
            "target": "_Trigger",
            "type": "lambda",
          },
        },
      ],
      "isGroup": true,
      "operator": undefined,
    },
    {
      "identifier": "SitnBaseTemplateID",
      "literal": "'FICA_MASSACT_MSG_MONITOR'",
      "operator": "eq",
    },
  ],
  "operator": "AND",
}
`;
